File: src_mcp_search/main/java/PanelTools.java
/*
 * Licensed to Crate.io GmbH ("Crate") under one or more contributor
 * license agreements.  See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.  Crate licenses
 * this file to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.  You may
 * obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 */

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Loads {@code grafana/german_weather_data.json} and turns every panel
 * with a {@code rawSql} target into a callable tool.
 *
 * <p>The Java port mirrors {@code german_weather_panels.py}: a panel's
 * title becomes a slugified tool name; its rawSql becomes the
 * template that gets substituted at call-time; Grafana template
 * variables ($var, ${var}) and the $__timeFilter() macro become tool
 * parameters with explicit JSON schema entries.
 */
public final class PanelTools {

    /** $var or ${var}, captures the name. */
    private static final Pattern TEMPLATE_VAR = Pattern.compile("\\$\\{?(\\w+)}?");

    /** $__timeFilter("column"), captures the column expression. */
    private static final Pattern TIME_FILTER = Pattern.compile("\\$__timeFilter\\(\\s*([^)]+?)\\s*\\)");

    /** Grafana template variable -> (param name, JSON-schema type, description). */
    private static final Map<String, ParamSpec> VAR_SPEC = new HashMap<>();
    static {
        VAR_SPEC.put("region", new ParamSpec("region", "string",
                "Federal-state name (one of the 16 German Länder), or the literal "
                        + "string 'ALL' for the whole country. There is no row named 'Germany'."));
        VAR_SPEC.put("keyword", new ParamSpec("keyword", "string", "Full-text search term"));
        VAR_SPEC.put("search_categories", new ParamSpec("search_categories", "string",
                "A single fulltext-indexed column on demo.german_regions to MATCH against. "
                        + "Valid values: 'region_name', 'economics', 'transportation', "
                        + "'introduced_species'. Pass exactly one column name — "
                        + "comma-separated lists produce a SQL syntax error, and 'tourism' "
                        + "is not fulltext-indexed."));
        VAR_SPEC.put("latitude", new ParamSpec("latitude", "number", "Latitude of the location"));
        VAR_SPEC.put("longitude", new ParamSpec("longitude", "number", "Longitude of the location"));
    }

    private PanelTools() {}

    /** Immutable tool definition exposed to Claude. */
    public static final class PanelTool {
        public final String name;
        public final String description;
        /** JSON-schema object suitable for {@code Tool.InputSchema.properties(...)}. */
        public final Map<String, Object> inputSchemaProperties;
        public final List<String> required;
        public final Function<Map<String, Object>, String> handler;

        PanelTool(String name,
                  String description,
                  Map<String, Object> inputSchemaProperties,
                  List<String> required,
                  Function<Map<String, Object>, String> handler) {
            this.name = name;
            this.description = description;
            this.inputSchemaProperties = inputSchemaProperties;
            this.required = required;
            this.handler = handler;
        }
    }

    private record ParamSpec(String paramName, String jsonType, String description) {}

    /**
     * Load the dashboard JSON and build one tool per panel-with-rawSql.
     */
    public static List<PanelTool> loadFromFile(Path dashboardPath, CrateDbClient crate) throws IOException {
        String text = Files.readString(dashboardPath);
        JsonObject root = JsonParser.parseString(text).getAsJsonObject();
        JsonArray panels = root.has("panels") ? root.getAsJsonArray("panels") : new JsonArray();

        List<PanelTool> tools = new ArrayList<>();
        Set<String> usedNames = new HashSet<>();
        for (JsonElement el : panels) {
            JsonObject panel = el.getAsJsonObject();
            if (panel.has("type") && "row".equals(panel.get("type").getAsString())) {
                continue; // row panels are section headers, no queries
            }
            PanelTool tool = makeTool(panel, crate, usedNames);
            if (tool != null) {
                tools.add(tool);
            }
        }
        return tools;
    }

    private static PanelTool makeTool(JsonObject panel, CrateDbClient crate, Set<String> usedNames) {
        if (!panel.has("targets")) {
            return null;
        }
        String sqlTemplate = null;
        for (JsonElement t : panel.getAsJsonArray("targets")) {
            JsonObject target = t.getAsJsonObject();
            if (target.has("rawSql")) {
                sqlTemplate = target.get("rawSql").getAsString();
                break;
            }
        }
        if (sqlTemplate == null) {
            return null;
        }

        String title = panel.has("title") ? panel.get("title").getAsString()
                                          : "panel_" + (panel.has("id") ? panel.get("id").getAsString() : "?");

        String name = slugify(title);
        if (usedNames.contains(name)) {
            name = name + "_" + (panel.has("id") ? panel.get("id").getAsString() : "x");
        }
        usedNames.add(name);

        Scan scan = scanSql(sqlTemplate);

        // Build properties + required list in stable (insertion) order.
        Map<String, Object> properties = new LinkedHashMap<>();
        List<String> required = new ArrayList<>();
        for (String var : scan.vars) {
            ParamSpec spec = VAR_SPEC.get(var);
            if (spec == null) {
                continue;
            }
            properties.put(spec.paramName, Map.of(
                    "type", spec.jsonType,
                    "description", spec.description));
            required.add(spec.paramName);
        }
        if (scan.usesTimeFilter) {
            properties.put("time_from", Map.of(
                    "type", "string",
                    "description", "Start of the time window, ISO 8601 (e.g. 2025-12-01)."));
            properties.put("time_to", Map.of(
                    "type", "string",
                    "description", "End of the time window, ISO 8601 (e.g. 2025-12-31)."));
            required.add("time_from");
            required.add("time_to");
        }

        String description = "Grafana panel '" + title + "'. Runs the panel's SQL against CrateDB and returns the rows.";

        // Capture in final locals for the lambda.
        final String capturedSql = sqlTemplate;
        final String capturedTitle = title;
        Function<Map<String, Object>, String> handler = args -> {
            try {
                String rendered = renderSql(capturedSql, args);
                JsonObject result = crate.runSql(rendered);
                return CrateDbClient.formatResult(rendered, result);
            } catch (Exception exc) {
                return "Error running panel '" + capturedTitle + "': " + exc.getMessage();
            }
        };

        return new PanelTool(name, description, properties, required, handler);
    }

    // ---- Slug / scan / render helpers ----

    /** Strip $/{} from template variable references then snake_case + lowercase. */
    static String slugify(String title) {
        String s = TEMPLATE_VAR.matcher(title).replaceAll("$1");
        s = s.replaceAll("[^A-Za-z0-9]+", "_").replaceAll("^_+|_+$", "").toLowerCase();
        return s.isEmpty() ? "panel" : s;
    }

    private static final class Scan {
        final Set<String> vars;
        final boolean usesTimeFilter;

        Scan(Set<String> vars, boolean usesTimeFilter) {
            this.vars = vars;
            this.usesTimeFilter = usesTimeFilter;
        }
    }

    /** Inspect SQL and report which template variables it uses + whether $__timeFilter appears. */
    static Scan scanSql(String sql) {
        Set<String> vars = new java.util.LinkedHashSet<>();
        Matcher m = TEMPLATE_VAR.matcher(sql);
        while (m.find()) {
            String name = m.group(1);
            if (!name.startsWith("__")) {
                vars.add(name);
            }
        }
        return new Scan(vars, sql.contains("$__timeFilter"));
    }

    /** Substitute caller arguments into the panel SQL template. */
    static String renderSql(String sql, Map<String, Object> args) {
        // Pass 1: $__timeFilter("col") -> (col >= 'time_from' AND col <= 'time_to')
        StringBuilder out = new StringBuilder();
        Matcher m = TIME_FILTER.matcher(sql);
        int idx = 0;
        while (m.find()) {
            out.append(sql, idx, m.start());
            String col = m.group(1).trim();
            Object tf = args.get("time_from");
            Object tt = args.get("time_to");
            if (tf == null || tt == null) {
                throw new IllegalArgumentException(
                        "time_from and time_to are required for panels that use $__timeFilter");
            }
            out.append('(').append(col).append(" >= '").append(tf)
               .append("' AND ").append(col).append(" <= '").append(tt).append("')");
            idx = m.end();
        }
        out.append(sql, idx, sql.length());
        String rendered = out.toString();

        // Pass 2: $var / ${var} -> arg value
        for (Map.Entry<String, ParamSpec> e : VAR_SPEC.entrySet()) {
            String var = e.getKey();
            Object value = args.get(e.getValue().paramName);
            if (value == null) {
                continue;
            }
            Pattern pat = Pattern.compile("\\$\\{?" + Pattern.quote(var) + "}?");
            rendered = pat.matcher(rendered).replaceAll(Matcher.quoteReplacement(value.toString()));
        }
        return rendered;
    }

    /**
     * Optional helper: locate the dashboard JSON relative to the
     * Java module's working directory. The Maven pom puts the source
     * dir at {@code src_mcp_search/main/java/}, so the dashboard sits
     * three levels up.
     */
    public static Path defaultDashboardPath() {
        Path here = Path.of("").toAbsolutePath();
        return here.resolve(Path.of("..", "..", "..", "grafana", "german_weather_data.json")).normalize();
    }

    public static Optional<Path> findDashboard() {
        // Walk up looking for a 'grafana/german_weather_data.json' until we hit the repo root.
        Path p = Path.of("").toAbsolutePath();
        for (int i = 0; i < 6; i++) {
            Path candidate = p.resolve("grafana").resolve("german_weather_data.json");
            if (Files.exists(candidate)) {
                return Optional.of(candidate);
            }
            Path parent = p.getParent();
            if (parent == null) {
                break;
            }
            p = parent;
        }
        return Optional.empty();
    }
}

File: src_mcp_search/main/dotnet/PanelTools.cs
// Licensed to Crate.io GmbH ("Crate") under one or more contributor
// license agreements.  See the NOTICE file distributed with this work for
// additional information regarding copyright ownership.  Crate licenses
// this file to you under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.  You may
// obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0

using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace GermanWeather;

/// <summary>
/// Loads <c>grafana/german_weather_data.json</c> and turns every panel
/// with a <c>rawSql</c> target into a callable tool. Mirrors the Java
/// PanelTools: panel title -> slugified tool name; rawSql -> template;
/// Grafana template variables ($var, ${var}) and the $__timeFilter()
/// macro -> tool parameters with JSON-schema entries.
/// </summary>
public static class PanelTools
{
    /// <summary>$var or ${var}; capture the bare name.</summary>
    private static readonly Regex TemplateVar = new(@"\$\{?(\w+)\}?", RegexOptions.Compiled);

    /// <summary>$__timeFilter("column"); capture the column expression.</summary>
    private static readonly Regex TimeFilter = new(@"\$__timeFilter\(\s*([^)]+?)\s*\)", RegexOptions.Compiled);

    private sealed record ParamSpec(string ParamName, string JsonType, string Description);

    private static readonly Dictionary<string, ParamSpec> VarSpec = new()
    {
        ["region"] = new ParamSpec("region", "string",
            "Federal-state name (one of the 16 German Länder), or the literal "
            + "string 'ALL' for the whole country. There is no row named 'Germany'."),
        ["keyword"] = new ParamSpec("keyword", "string", "Full-text search term"),
        ["search_categories"] = new ParamSpec("search_categories", "string",
            "A single fulltext-indexed column on demo.german_regions to MATCH against. "
            + "Valid values: 'region_name', 'economics', 'transportation', "
            + "'introduced_species'. Pass exactly one column name — "
            + "comma-separated lists produce a SQL syntax error, and 'tourism' "
            + "is not fulltext-indexed."),
        ["latitude"] = new ParamSpec("latitude", "number", "Latitude of the location"),
        ["longitude"] = new ParamSpec("longitude", "number", "Longitude of the location"),
    };

    /// <summary>Immutable tool definition exposed to Claude.</summary>
    public sealed class PanelTool
    {
        public string Name { get; }
        public string Description { get; }
        /// <summary>JSON-schema object suitable for the Anthropic tool input_schema.properties.</summary>
        public Dictionary<string, object> InputSchemaProperties { get; }
        public List<string> Required { get; }
        public Func<Dictionary<string, object?>, Task<string>> Handler { get; }

        public PanelTool(
            string name,
            string description,
            Dictionary<string, object> inputSchemaProperties,
            List<string> required,
            Func<Dictionary<string, object?>, Task<string>> handler)
        {
            Name = name;
            Description = description;
            InputSchemaProperties = inputSchemaProperties;
            Required = required;
            Handler = handler;
        }
    }

    public static async Task<List<PanelTool>> LoadFromFileAsync(string dashboardPath, CrateDbClient crate)
    {
        var text = await File.ReadAllTextAsync(dashboardPath).ConfigureAwait(false);
        var root = JsonNode.Parse(text) as JsonObject ?? throw new InvalidOperationException("dashboard root is not a JSON object");
        var panels = root["panels"] as JsonArray ?? new JsonArray();

        var tools = new List<PanelTool>();
        var usedNames = new HashSet<string>();
        foreach (var el in panels)
        {
            if (el is not JsonObject panel) continue;
            if (panel["type"]?.GetValue<string>() == "row") continue;
            var tool = MakeTool(panel, crate, usedNames);
            if (tool != null) tools.Add(tool);
        }
        return tools;
    }

    private static PanelTool? MakeTool(JsonObject panel, CrateDbClient crate, HashSet<string> usedNames)
    {
        if (panel["targets"] is not JsonArray targets) return null;

        string? sqlTemplate = null;
        foreach (var t in targets)
        {
            if (t is JsonObject target && target["rawSql"] is JsonNode raw)
            {
                sqlTemplate = raw.GetValue<string>();
                break;
            }
        }
        if (sqlTemplate == null) return null;

        var title = panel["title"]?.GetValue<string>()
            ?? $"panel_{panel["id"]?.ToString() ?? "?"}";

        var name = Slugify(title);
        if (usedNames.Contains(name))
        {
            name = $"{name}_{panel["id"]?.ToString() ?? "x"}";
        }
        usedNames.Add(name);

        var (vars, usesTimeFilter) = ScanSql(sqlTemplate);

        var properties = new Dictionary<string, object>();
        var required = new List<string>();
        foreach (var v in vars)
        {
            if (!VarSpec.TryGetValue(v, out var spec)) continue;
            properties[spec.ParamName] = new Dictionary<string, object>
            {
                ["type"] = spec.JsonType,
                ["description"] = spec.Description,
            };
            required.Add(spec.ParamName);
        }
        if (usesTimeFilter)
        {
            properties["time_from"] = new Dictionary<string, object>
            {
                ["type"] = "string",
                ["description"] = "Start of the time window, ISO 8601 (e.g. 2025-12-01).",
            };
            properties["time_to"] = new Dictionary<string, object>
            {
                ["type"] = "string",
                ["description"] = "End of the time window, ISO 8601 (e.g. 2025-12-31).",
            };
            required.Add("time_from");
            required.Add("time_to");
        }

        var description = $"Grafana panel '{title}'. Runs the panel's SQL against CrateDB and returns the rows.";

        var capturedSql = sqlTemplate;
        var capturedTitle = title;
        async Task<string> Handler(Dictionary<string, object?> args)
        {
            try
            {
                var rendered = RenderSql(capturedSql, args);
                var result = await crate.RunSqlAsync(rendered).ConfigureAwait(false);
                return CrateDbClient.FormatResult(rendered, result);
            }
            catch (Exception exc)
            {
                return $"Error running panel '{capturedTitle}': {exc.Message}";
            }
        }

        return new PanelTool(name, description, properties, required, Handler);
    }

    /// <summary>Strip $/{} from template variable references, then snake_case + lowercase.</summary>
    internal static string Slugify(string title)
    {
        var s = TemplateVar.Replace(title, "$1");
        s = Regex.Replace(s, "[^A-Za-z0-9]+", "_");
        s = Regex.Replace(s, "^_+|_+$", "");
        s = s.ToLowerInvariant();
        return string.IsNullOrEmpty(s) ? "panel" : s;
    }

    internal static (List<string> Vars, bool UsesTimeFilter) ScanSql(string sql)
    {
        var seen = new HashSet<string>();
        var vars = new List<string>();
        foreach (Match m in TemplateVar.Matches(sql))
        {
            var name = m.Groups[1].Value;
            if (name.StartsWith("__")) continue;
            if (seen.Add(name)) vars.Add(name);
        }
        return (vars, sql.Contains("$__timeFilter"));
    }

    internal static string RenderSql(string sql, Dictionary<string, object?> args)
    {
        // Pass 1: $__timeFilter("col") -> (col >= 'time_from' AND col <= 'time_to')
        var rendered = TimeFilter.Replace(sql, m =>
        {
            var col = m.Groups[1].Value.Trim();
            if (!args.TryGetValue("time_from", out var tf) || tf == null ||
                !args.TryGetValue("time_to", out var tt) || tt == null)
            {
                throw new ArgumentException(
                    "time_from and time_to are required for panels that use $__timeFilter");
            }
            return $"({col} >= '{tf}' AND {col} <= '{tt}')";
        });

        // Pass 2: $var / ${var} -> arg value
        foreach (var (var, spec) in VarSpec)
        {
            if (!args.TryGetValue(spec.ParamName, out var value) || value == null) continue;
            var pat = new Regex(@"\$\{?" + Regex.Escape(var) + @"\}?");
            rendered = pat.Replace(rendered, value.ToString() ?? "");
        }
        return rendered;
    }

    /// <summary>
    /// Walk up from cwd looking for <c>grafana/german_weather_data.json</c>.
    /// </summary>
    public static string? FindDashboard()
    {
        var p = new DirectoryInfo(Directory.GetCurrentDirectory());
        for (var i = 0; i < 6 && p != null; i++)
        {
            var candidate = Path.Combine(p.FullName, "grafana", "german_weather_data.json");
            if (File.Exists(candidate)) return candidate;
            p = p.Parent;
        }
        return null;
    }
}
